### ISSUE TITLE
SBT often recompiles lots of files or complains about not found symbols

### ISSUE THREAD

**Issue by nafg** [sbt/sbt#6689]
## steps

I often have 2 SBT processes, one continuously recompiling JVM modules and one continuously recompiling JS modules. There are many shared modules that they depend on directly and transitively.

I automate this development loop with the following script:

```bash
#!/usr/bin/env bash

tmux new -d "docker-compose up db"

tmux split -h "$SHELL"

tmux split -f -p 80 "sbt ~app_lrbcol/reStart"

tmux split -h "sleep 5; sbt ~app_js/fastOptJS/webpack"

tmux attach
```

## problem

My feedback cycle is ridiculously long. Very often I get tons of errors about legitimate symbols not being found. On the next compile, sometimes triggered immediately automatically and sometimes triggered manually (e.g. by pressing R), many modules are recompiled. There's no legitimate reason for this, and it makes my development experience horrible. This has been going on for a very long time, months or years.

Here is are some typical errors:

`[error] /home/naftoli/dev/gitlab.com/chesednow/disp/people_views_shared/src/main/scala/chesednow/people/views/shared/PersonRoutePaths.scala:48:24: not found: type LLookup                                           `

`[error] /home/naftoli/dev/gitlab.com/chesednow/disp/custom_fields_shared/src/main/scala/chesednow/customfields/shared/FoldLookupType.scala:8:18: object equipment is not a member of package chesednow               `

Of course, there are tons of secondary errors caused by code that should be valid not being valid.

Here is some output of it recompiling way more than necessary:

```
[info] Build triggered by /home/naftoli/dev/gitlab.com/chesednow/disp/request_js/src/main/scala/chesednow/requests/sjs/list/Row.scala. Running 'app_js/fastOptJS/webpack'.
[info] compiling 25 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/models_shared/.js/target/classes ...
[info] compiling 2 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/delivery_shared/.js/target/classes ...
[info] compiling 3 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/volunteer_shared/.js/target/classes ...
[info] compiling 30 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/js_common/target/classes ...
[info] compiling 20 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/equipment_shared/.js/target/classes ...
[info] compiling 2 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/notes_shared/.js/target/classes ...
[info] compiling 5 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/people_models_shared/.js/target/classes ...
[info] compiling 20 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/doctors_shared/.js/target/classes ...
[info] compiling 2 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/notes_js/target/classes ...
[info] compiling 34 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/request_shared/.js/target/classes ...
[info] compiling 6 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/people_views_shared/.js/target/classes ...
[info] compiling 2 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/equipment_views_shared/.js/target/classes ...
[info] compiling 11 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/custom_fields_shared/.js/target/classes ...
[info] compiling 38 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/people_js/target/classes ...
[info] compiling 9 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/request_views_shared/.js/target/classes ...
[error] error while loading NameSuffix, /home/naftoli/dev/gitlab.com/chesednow/disp/doctors_shared/.js/target/classes/chesednow/doctors/shared/NameSuffix.class (No such file or directory)
[error] error while loading DoctorContact, /home/naftoli/dev/gitlab.com/chesednow/disp/doctors_shared/.js/target/classes/chesednow/doctors/shared/DoctorContact.class (No such file or directory)
[error] two errors found
[error] (people_js / Compile / compileIncremental) Compilation failed
[error] Total time: 233 s (03:53), completed Oct 13, 2021, 11:33:02 PM
[info] 21. Monitoring source files for root/app_js/fastOptJS/webpack...
[info]     Press <enter> to interrupt or '?' for more options.
[info] Received input event: Trigger.
[info] compiling 25 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/models_shared/.js/target/classes ...
[info] compiling 2 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/delivery_shared/.js/target/classes ...
[info] compiling 3 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/volunteer_shared/.js/target/classes ...
[info] compiling 30 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/js_common/target/classes ...
[info] compiling 20 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/equipment_shared/.js/target/classes ...
[info] compiling 2 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/notes_shared/.js/target/classes ...
[info] compiling 5 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/people_models_shared/.js/target/classes ...
[info] compiling 20 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/doctors_shared/.js/target/classes ...
[info] compiling 2 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/notes_js/target/classes ...
[info] compiling 34 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/request_shared/.js/target/classes ...
[info] compiling 6 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/people_views_shared/.js/target/classes ...
[info] compiling 2 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/equipment_views_shared/.js/target/classes ...
[info] compiling 11 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/custom_fields_shared/.js/target/classes ...
[info] compiling 38 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/people_js/target/classes ...
[info] compiling 41 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/request_js/target/classes ...
[info] compiling 14 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/delivery_js/target/classes ...
[info] compiling 15 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/equipment_js/target/classes ...
[info] compiling 19 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/doctors_js/target/classes ...
[info] compiling 4 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/app_js/target/classes ...
[info] Fast optimizing /home/naftoli/dev/gitlab.com/chesednow/disp/app_js/target/app_js-fastopt
[success] Total time: 218 s (03:38), completed Oct 13, 2021, 11:36:59 PM
[info] 22. Monitoring source files for root/app_js/fastOptJS/webpack...
[info]     Press <enter> to interrupt or '?' for more options.
```

## expectation

Compiling shared modules to one target should not step on the toes of compiling them to another target. Alternatively, it should be possible to tell one SBT processes to do two separate continuous tasks in parallel.

**Comment by nafg** [comment]
Some notes: 

1. It does not only happen when both processes are compiling simultaneously. Even if only one, say the JVM compilation, is active, and the other one is e.g. sitting at "Monitoring source files ..." and not doing anything, it often takes numerous attempts until compilation (of JVM outputs in this example) succeeds. Even then, sometimes the application starts and crashes with a ClassNotFoundException (although that's much less common).

2. I mean this less as a bug report ("here is something for you to fix") and more as a request for help ("how on earth can I debug this and figure out what's causing it"), although of course if the conclusion is that it's a limitation of SBT then removing that limitation somehow would be nice (but I could potentially switch to Mill).

3. It's very unfortunate that there's no way to do in a single sbt shell both `~app_lrbcol/reStart` and `~app_js/fastOptJS/webpack` concurrently. If it were possible then I could also do it in separate terminal panes via `sbtn`, but it's not possible to do separate `~` commands in separate `sbtn` invocations.

**Comment by eed3si9n** [comment]
We don't support multiple instances of sbt. Why not create a task or command that compose the two things and run that continuously instead?

**Comment by nafg** [comment]
Because the frontend and the backend have a different definition of "continuously."

 * When frontend code changes, I don't want `reStart` to run. I don't want the backend to be shut down and then have to wait for it to finish booting up again.
 * Off the top of my head I don't see any harm in running fastOptJS/webpack when backend code changes. It might waste a small amount of time, or it might be instantaneous, either way not the end of the world. To reload frontend code in this project I currently use [Live.js](https://livejs.com/), which is not very efficient. If I were using something like webpack dev server, I'm not sure if it would cause any issues. Probably not, I suppose. I guess only if there was a command that needed to be run to tell it to reload, but usually it's just detected if files change. So perhaps in this direction there's no issue. I guess it's mainly the other way, that I need to avoid reloading the backend when frontend code changes. Unless it just ends up being too much slower for SBT to run through the no-op frontend tasks every time it's triggered by backend code changes.

**Comment by nafg** [comment]
In another project, IIRC I tried to implement this in build code -- the backend had a task that was supposed to only run `reStart` if it was triggered by something in the backend changing. That way a single command could continuously trigger only whatever would be appropriate to trigger. I don't think I managed to figure out how to do that correctly and reliably, though.

There's another thing also.

I find it's a much better DX to have output from the backend `reStart` loop separate from the output from the frontend compilation loop (e.g. in separate terminal panes). For instance, if I enable more verbose logging on the backend, there could be a lot of output that will drown out compilation errors when I make frontend code changes.

Now, if a single SBT instance could have multiple `~` loops, then I could solve that by running separate `sbtn` commands. Alternatively, there might be other ways to hack it. For example, the application could log to a file and there could be a separate terminal pane that `tail`s the file. (But then I would lose the ability to see colored logging.)

**Comment by nafg** [comment]
I don't think it really has to do with separate sbt instances. I'm running with a single instance now and it's behaving similarly.

Anyway, we could make guesses about what the issue might be, or talk about expectations (as in should I expect this attempt to work), but what I really want to do is figure out the cause.

I would like to be able to see, without turning on torrents of logs, exactly what is triggering each compilation step. For example this doesn't help:

```
[info] Build triggered by /home/naftoli/dev/gitlab.com/chesednow/disp/request_js/src/main/scala/chesednow/requests/sjs/list/Filters.scala. Running 'app_lrbcol/reStart'.
[info] compiling 25 Scala sources to /home/naftoli/dev/gitlab.com/chesednow/disp/models_shared/.jvm/target/classes ...
````

`models_shared` does not depend on `request_js` (the reverse is true, transitively). SBT is telling me why it did the thing after `~`, which is nice, but what I need to know is why it thinks 25 files in `models_shared` are invalidated.

**Comment by eed3si9n** [comment]
Maybe try `backend/bgRun` and see if that works. It's kind of like `reStart`.

**Comment by nafg** [comment]
Same

**Comment by nafg** [comment]
Tangentially, is there any way to stop a `bgRun` process? I keep getting `java.net.BindException: Address already in use`.

**Comment by eed3si9n** [comment]
bgList would list and bgStop should stop.

In any case, if this is an incremental compilation issue, you should be able to reproduce this just with `~app_js/fastOptJS/webpack`. I added some comments in `Row.scala`, but was not reproduce overcompilation.
I also ran `~app_lrbcol/Compile/products` in an (unsupported) second instance, but didn't change the situation.

If there's a Zinc issue, debug log is really your tool.

**Comment by nafg** [comment]
So how do I turn on Zinc debug logs? And is it going to print 100X the output I need (as I recall from trying it a long time ago) or is it possible to get more manageable output?

**Comment by eed3si9n** [comment]
I usually type

```
debug
```

into the shell to see all the logs. You will see a ton of logs, but hopefully earlier in the log you should see the invalidation event that's causing the problem.

**Comment by nafg** [comment]
Here is a bunch of output. I have no idea what to learn from it.

https://gist.github.com/nafg/a7974cb45cc2448605bce5c6521bf04e

**Comment by eed3si9n** [comment]
### all 3 sources are invalidated


```
[debug] [zinc] IncrementalCompile -----------
[debug] IncrementalCompile.incrementalCompile
[debug] previous = Stamps for: 42 products, 3 sources, 8 libraries
[debug] current source = Set(${BASE}/volunteer_shared/src/main/scala/chesednow/volunteers/shared/Vehicle.scala, ${BASE}/volunteer_shared/src/main/scala/chesednow/volunteers/shared/VolunteerGroup.scala, ${BASE}/volunteer_shared/src/main/scala/chesednow/volunteers/shared/VolunteerInfo.scala)
[debug] [zinc] IncrementalCompile -----------
[debug] IncrementalCompile.incrementalCompile
[debug] previous = Stamps for: 351 products, 20 sources, 16 libraries
[debug] current source = Set(${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentCategoriesApiModule.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentCategory.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentLocation.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentItemCode.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentSelection.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentItem.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentDispensingRestock.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentType.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentEntryPoints.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentItemCheckIn.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentTree.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentDispensingTypesApiModule.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentApiModule.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentLabelsApiModule.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentItemInfo.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentCondition.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/Availability.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentLabel.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentLoanInfo.scala, ${BASE}/equipment_shared/src/main/scala/chesednow/equipment/shared/EquipmentLoanTypesApiModule.scala)
[debug] > initialChanges = InitialChanges(Changes(added = Set(), removed = Set(), changed = Set(), unmodified = ...),Set(${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$$anon$3.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$anon$lazy$macro$9$4.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$anon$lazy$macro$3$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$anon$lazy$macro$9$3$$anon$9.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$$anon$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$anon$lazy$macro$7$1$$anon$3.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$anon$lazy$macro$7$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$$anon$8.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$anon$lazy$macro$9$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$anon$lazy$macro$7$2$$anon$4.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$anon$lazy$macro$9$3.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfoLenses.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$$anon$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$$anon$6.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$anon$lazy$macro$3$1$$anon$11.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$$anon$3.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$anon$lazy$macro$7$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$anon$lazy$macro$3$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$anon$lazy$macro$3$2$$anon$12.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$anon$lazy$macro$9$1$$anon$4.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$anon$lazy$macro$9$2$$anon$5.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$anon$lazy$macro$9$1$$anon$4.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$anon$lazy$macro$9$2$$anon$5.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$$anon$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$$anon$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$anon$lazy$macro$9$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$$anon$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$anon$lazy$macro$9$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$$anon$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$anon$lazy$macro$9$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$anon$lazy$macro$9$4$$anon$10.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$$anon$7.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$.class),Set(),API Changes: Set())
[debug] 
[debug] Initial source changes:
[debug] 	removed: Set()
[debug] 	added: Set()
[debug] 	modified: Set()
[debug] Invalidated products: Set(${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$$anon$3.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$anon$lazy$macro$9$4.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$anon$lazy$macro$3$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$anon$lazy$macro$9$3$$anon$9.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$$anon$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$anon$lazy$macro$7$1$$anon$3.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$anon$lazy$macro$7$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$$anon$8.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$anon$lazy$macro$9$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$anon$lazy$macro$7$2$$anon$4.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$anon$lazy$macro$9$3.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfoLenses.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$$anon$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$$anon$6.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$anon$lazy$macro$3$1$$anon$11.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$$anon$3.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$anon$lazy$macro$7$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$anon$lazy$macro$3$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$anon$lazy$macro$3$2$$anon$12.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$anon$lazy$macro$9$1$$anon$4.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$anon$lazy$macro$9$2$$anon$5.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$anon$lazy$macro$9$1$$anon$4.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$anon$lazy$macro$9$2$$anon$5.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$$anon$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$$anon$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$anon$lazy$macro$9$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerGroup$$anon$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/Vehicle$anon$lazy$macro$9$1.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$$anon$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$anon$lazy$macro$9$2.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$anon$lazy$macro$9$4$$anon$10.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$Existing$$anon$7.class, ${BASE}/volunteer_shared/.jvm/target/classes/chesednow/volunteers/shared/VolunteerInfo$New$.class)
[debug] External API changes: API Changes: Set()
[debug] Modified binary dependencies: Set()
[debug] Initial directly invalidated classes: Set()
[debug] Sources indirectly invalidated by:
[debug] 	product: Set(${BASE}/volunteer_shared/src/main/scala/chesednow/volunteers/shared/VolunteerGroup.scala, ${BASE}/volunteer_shared/src/main/scala/chesednow/volunteers/shared/VolunteerInfo.scala, ${BASE}/volunteer_shared/src/main/scala/chesednow/volunteers/shared/Vehicle.scala)
[debug] 	binary dep: Set()
[debug] 	external source: Set()
[debug] all 3 sources are invalidated
```

Just kind of looking at the log, it seems like the invalidation is happening as full invalidation, as in all sources in the subprojects are invalidated. Did you change any of these sources?

For now let's assume you did. This is triggering full build in other parts of the code either because of the macros and implicits used in those sources, or macros in other sources that now need to be reevaluated due to the changes in macros. Typeclass instances are like root of the proof evidence. Once you change any of them, it's not surprising that would trigger a full build.

**Comment by nafg** [comment]
I did not modify any of those files.

**Comment by nafg** [comment]
I think it's happening with Mill too, though I'm not certain. So what is going on? Is some program on my computer writing to random files??

What triggers SBT or Zinc to invalidate a file? If its modification timestamp changes it will cause downstream invalidation even if its contents or the same? Or does it matter if its contents change? Or is it even just inotify events? Could it be that it's reacting to something it shouldn't be?

And what could be doing it? Could it be IntelliJ? I never heard of that happening for anyone else. Anyone know how I can see what programs are causing file events (on Ubuntu)? I think I went down this rabbit hole once and didn't get anywhere but I don't remember the specifics.

**Comment by eed3si9n** [comment]
I think I switch to using content hash for normal compilation, but during `~` inotify might trigger something.
Also back to my original statement, sbt is not designed to run multiple instances at the same time, so really all bets are off in terms of correct behavior. One instance could be saving some temporary information, and the other instance could be reading from it and doing something odd etc.

**Comment by nafg** [comment]
Again, it happens even with only a single instance

**Comment by eed3si9n** [comment]
Since it would be difficult to change the operating system, I guess you could try coding without IntelliJ for a day, and then coding with IntelliJ but without `~` next day and to narrow down the parameters?

**Comment by nafg** [comment]
I'm not very productive without IntelliJ. Instead what I did is I ported my build to Mill (I have some code in both builds that read dependencies from a HOCON file so I can maintain the build in SBT and Mill). I did experience this sort of recompile once but that's it. Maybe I'll give SBT another try one of these days and see if it still has the issue. Maybe I should try running SBT in the terminal while coding in an IntelliJ project that's based on the Mill build, in case the issue was related to IntelliJ stepping on SBT's toes somehow.
